rfcs/2022-11-11-smp-basic-auth.md

```diff
@@ -0,0 +1,27 @@
+# SMP Basic Auth
+
+## Problem
+
+Users who host their own servers do not want unknown people to be able to create messaging queues on their servers after discovering server address in groups or after making a connection. As the number of self-hosted servers is growing it became more important than it was when we excluded it from the original design.
+
+## Solution
+
+Single access password that can be optionally included in server address that is passed to app configuration. It will not be allowed in the existing contexts (and parsing will fail), to avoid accidentally leaking it. Server address with password will look this way: `smp://fingerprint:password@hosts`
+
+## Implementation plan
+
+1. A separate type to include server and password, so it can only be used where allowed.
+
+2. Server password to create queues will be configured in TRANSPORT section of INI file, as `create_password` parameter.
+
+3. The password will only be required in server configuration/address to create queues only, it won't be required for other receiving queue operations on already existing queues.
+
+4. If new command is attempted in the session that does not allow creating queues, the server will send `ERR AUTH` response
+
+5. Passing password to the server can be done in one of the several ways, we need to decide:
+
+  - as a parameter of NEW command. Pros: a local change, that only needs checking when queue is created. Cons: protocol version change.
+  - as a separate command AUTH. Pros: allows to include additional parameters and potentially be extended beyond basic auth. Cons: more complex to manage server state, can be more difficult syntax in the future, if extended.
+  - as part of handshake (we currently ignore the unparsed part of handshake block, so it can be extended). Pros: probably, the simplest, and independent of the commands protocol – establishes create permission for the current session. Cons: the client won't know about whether it is able to create the queue until it tries (same as in case 1).
+
+  My preference is the last option. As a variant of the last option, we can add a server response/message that includes permission to create queues - it will only be sent to the clients who pass credential in handshake - that might simplify testing server connection (we currently do not do it). It might be unnecessary, as we could simply create and delete queue in case credential is passed as part of testing connection (and even sending a message to it).
```

Comment:
> ... we could simply create and delete queue in case credential is passed as part of testing connection (and even sending a message to it).

even though we do treat queues as inexpensive, this still seems wasteful..

Reply:
we don't do this now, I don't think there's greatly more reason to do it for authenticated connection if the implementation is correct.. client do retries or report error so it's also not that big of a problem - but if to solve it better via confirmation from server in the initial handshake / answer to AUTH

Reply:
I think there is actually a value to perform the actual test of some basic operations, as part of agent “test server” function - more than from simply informing the client that it works

Reply:
And as result tell the user which operations succeeded and which failed, and then delete the queue as part of the test

Reply:
It can be used to pre-flight servers before changing config, because if you change config and some server doesn’t work you may permanently fail some things - like receiving files etc.

Reply:
it's a good thing to do on config on change, but probably not on each new session